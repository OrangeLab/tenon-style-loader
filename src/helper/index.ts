export const defaultInclude = ['less', 'styl', 'scss']

export type Loader = string

export const hasTenonLoaderReg = /tenon-loader/
export const tenonStyleLoaderPath = require.resolve('../index')
export function isArray(array: any){
  return Object.prototype.toString.call(array) === '[object Array]'
}

export function isObject(object:any){
  return Object.prototype.toString.call(object) === '[object Object]'
}

export function isAutoRule(rule:RegExp, includes: Array<string>){
  return includes.some((postfix:string) => {
    return rule.test(`index.${postfix}`)
  })
}

export function appendLoader(rule:any):Array<Loader>{
  if(!rule){
    return []
  }
  let loaders:Array<Loader> = [tenonStyleLoaderPath]
  if(isObject(rule.use)){
    loaders.push(rule.use.loader)
    return loaders
  }
  if(isArray(rule.use)){
    loaders = loaders.concat(rule.use)
  }
  return loaders
}

export function notHasTenonLoader(rule:any):Boolean{
  if(isObject(rule.use)){
    return hasTenonLoaderReg.test(rule.use.loader)
  }
  if(isArray(rule.use)){
    return rule.use.some((loader:string) => {
      return hasTenonLoaderReg.test(loader)
    })
  }
  return false
}