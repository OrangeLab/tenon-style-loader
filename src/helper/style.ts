import postcss, { Root, Rule } from "postcss";

import { styleTransformer } from "@hummer/tenon-utils";
import {
  RuleSetMap,
  MatchType,
  CompileStyleOptions,
  RuleSet,
  ClassRule,
  Rule as CssRule,
  Selector,
} from "../style";

const isClassSelectorReg = /^\./;
const isTagSelectorReg = /\[.+\]/;
const isAttrSelectorReg = /\[.+\]/;
const isIDSelectorReg = /^\#/;
const isCommaSelectorReg = /,/;

/**
 * 针对Class的Selector进行特殊处理
 * 支持的Selectors如下：
 * 基础选择器：
 * #id ID选择器
 * .class Class选择器
 * view Tag选择器
 * 复杂选择器：属性选择器
 * .class[attr] 属性选择器
 * @param selector
 */
function handleSelector(
  ruleSetMap: RuleSetMap,
  selector: string,
  node: Rule,
  options: CompileStyleOptions
) {
  let selectorList = selector.split(/[\s,]/).filter((item) => !!item);
  let lastSelector = selectorList.pop() as string;
  let style = getRuleStyle(node);
  let { scoped, id = "" } = options;

  if (isClassSelectorReg.test(lastSelector)) {
    let className = lastSelector.slice(1);
    let selector = new Selector(className, MatchType.Class);
    let classRule = new ClassRule(selector, style);
    collectRuleWithScoped(ruleSetMap, MatchType.Class, classRule, scoped, id);
    return;
  }
  // TODO Support Attr Selector
  if (isAttrSelectorReg.test(lastSelector)) {
  }

  // TODO Support ID Selector
  if (isIDSelectorReg.test(lastSelector)) {
  }

  // TODO Support Tag Selector
  if (isTagSelectorReg.test(lastSelector)) {
  }
  return "";
}

function generateCode(ruleSetMap: RuleSetMap, options: CompileStyleOptions) {
  let { packageName } = options;
  let styleCode = `
    var ruleSetMap = ${ruleSetMap.toJsonString()};
    var options = ${JSON.stringify(options)};
  `;
  return `
    import {collectStyle} from '${packageName}';
    export default (function(){
      ${styleCode}
      return collectStyle(ruleSetMap, options);
    })();
  `;
}

function getRuleStyle(node: Rule): Record<string, string> {
  let style: any = {};
  node.each((item: any) => {
    let { prop, value } = item;
    style[prop] = value;
  });
  style = styleTransformer.transformStyle(style);
  return style;
}

export const compileStyle = function (
  source: string,
  options: CompileStyleOptions = { scoped: false }
) {
  let ruleSetMap = new RuleSetMap();
  postcss([getCollectPlugin(ruleSetMap, options)]).process(source, {
    from: undefined,
  }).css;
  let code = generateCode(ruleSetMap, options);
  return code;
};

export function getCollectPlugin(
  ruleSetMap: RuleSetMap,
  customOptions: CompileStyleOptions
) {
  return {
    postcssPlugin: "collect-rule",
    Once(root: Root) {
      root.each(function collectRule(node) {
        if (node.type !== "rule") {
          // 不支持媒体查询
          return;
        }
        let { selector } = node;
        if (isCommaSelector(selector)) {
          // Handle ',' case
          let selectors = selector.split(/,/).filter((item) => !!item);
          selectors.forEach((item) => {
            handleSelector(ruleSetMap, item, node, customOptions);
          });
        } else {
          handleSelector(ruleSetMap, selector, node, customOptions);
        }
      });
    },
  };
}

function isCommaSelector(selector: string) {
  return isCommaSelectorReg.test(selector);
}

/**
 * 处理样式隔离
 */
function collectRuleWithScoped(
  ruleSetMap: RuleSetMap,
  type: MatchType,
  rule: CssRule,
  scoped: boolean,
  id?: string
) {
  if (scoped && id) {
    // Support Scoped Css
    let ruleSet = ruleSetMap.getItem(id);
    if (!ruleSet) {
      ruleSet = new RuleSet();
      ruleSetMap.setItem(id, ruleSet);
    }
    ruleSet.appendRule(type, rule);
  } else {
    ruleSetMap.global.appendRule(type, rule);
  }
}
