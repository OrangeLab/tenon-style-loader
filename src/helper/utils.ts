import qs from 'querystring'

export function getOptions(loaderContext:any){
  const query = loaderContext.query;

  if (typeof query === 'string' && query !== '') {
    return qs.parse(loaderContext.query.slice(1));
  }

  if (!query || typeof query !== 'object') {
    return {};
  }

  return query;
}