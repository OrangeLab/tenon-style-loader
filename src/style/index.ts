export * from "./style-rule";

export interface CompileStyleOptions {
  scoped: boolean;
  id?: string;
  packageName?: string;
}
