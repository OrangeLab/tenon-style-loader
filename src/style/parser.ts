import { Selector, MatchType} from "."

enum Token {
  Dot = '.',
  Space = ' ',
  LeftBrace = '[',
  RightBrace = ']'
}

export function parseSelector(selector: string):(Selector | null){
  if(!selector){
    return null
  }

  let tokenStack = new TokenStack()
  for(let len=selector.length,i = len-1;i >= 0; i--){
    let curToken = selector.charAt(i)
    switch(curToken){
      case Token.Space:
        // TODO
        break;
      case Token.Dot:
        // TODO
        break;
      default:
        tokenStack.push(curToken)
        break;
    }
  }
  return new Selector(selector, MatchType.Class)
}



class Stack<T>{
  private _arr:Array<T> = []
  constructor(){}

  get arr(){
    return this._arr
  }

  push(item:T){
    this._arr.push(item)
  }

  pop(){
    return this._arr.pop()
  }

  empty(){
    this._arr = []
  }
}

class TokenStack extends Stack<string>{
  constructor(){
    super()
  }

  toToken(){
    return this.arr.reverse().reduce((last:string, current:string) => {
      return last + "" + current
    })
  }
}
