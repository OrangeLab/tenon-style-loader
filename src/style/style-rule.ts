import { StyleRelation } from "./style-relation";

export type Selector = string;
export interface Style {
  [key: string]: string | number;
}

export enum MatchType {
  Class,
  ID,
  Attr,
  Tag,
}

export class RuleSetMap {
  public global: RuleSet = new RuleSet();
  private scopedItems: Record<string, RuleSet> = {};
  constructor() {}

  getItem(id: string) {
    return this.scopedItems[id];
  }

  setItem(id: string, ruleSet: RuleSet) {
    this.scopedItems[id] = ruleSet;
  }

  toJsonString() {
    return JSON.stringify({
      global: this.global,
      ...this.scopedItems,
    });
  }
}

export class RuleSet {
  // tagList: Array<Rule> = [];
  classList: Array<Rule> = []; // 暂时只支持 Class Selector
  // idList: Array<Rule> = [];
  // attrList: Array<Rule> = [];

  constructor() {}

  appendRule(type: MatchType, rule: Rule) {
    switch (type) {
      case MatchType.Class:
        this.appendClassRule(rule);
        break;
        // case MatchType.Tag:
        //   this.appendTagRule(rule);
        //   break;
        // case MatchType.ID:
        //   this.appendIdRule(rule);
        //   break;
        // case MatchType.Attr:
        //   this.appendAttrRule(rule);
        break;
      default:
        break;
    }
  }

  appendClassRule(rule: ClassRule) {
    this.classList.push(rule);
  }

  // appendTagRule(rule: TagRule) {
  //   this.tagList.push(rule);
  // }

  // appendIdRule(rule: IdRule) {
  //   this.idList.push(rule);
  // }

  // appendAttrRule(rule: AttrRule) {
  //   this.classList.push(rule);
  // }
}

export class Rule {
  public selector: Selector;
  public matchType: MatchType;
  public relation: StyleRelation | null = null;
  public style: Style;
  constructor(selector: Selector, style: Style) {
    this.selector = selector;
    this.style = style;
  }

  appendRelation(relation: StyleRelation) {
    this.relation = relation;
  }
}

export class ClassRule extends Rule {
  matchType = MatchType.Class;
  constructor(selector: Selector, style: Style) {
    super(selector, style);
  }
}

export class TagRule extends Rule {
  matchType = MatchType.Tag;
  constructor(selector: Selector, style: Style) {
    super(selector, style);
  }
}

export class IdRule extends Rule {
  matchType = MatchType.ID;
  constructor(selector: Selector, style: Style) {
    super(selector, style);
  }
}

export class AttrRule extends Rule {
  matchType = MatchType.Attr;
  constructor(selector: Selector, style: Style) {
    super(selector, style);
  }
}
