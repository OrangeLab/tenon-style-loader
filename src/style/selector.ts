import { MatchType } from "./style-rule";

export enum RelationType {
  Subselector, // No combinator(Default) .a | .a.b
  DescendantSpace, // .a .b
  Child, // .a > .b
  DirectAdjacent, // .a + .b
  IndirectAdjacent, // .a ~ .b
}

export class Selector {
  value: string;
  matchType: MatchType;
  relation: RelationType;
  next: Selector | null = null;
  constructor(
    value: string,
    matchType: MatchType,
    relation: RelationType = RelationType.Subselector
  ) {
    this.value = value;
    this.matchType = matchType;
    this.relation = relation;
  }
  appendNext(selector: Selector) {
    this.next = selector;
  }
}

