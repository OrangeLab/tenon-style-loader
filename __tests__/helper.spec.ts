import { compileStyle } from "../src/helper/style";

describe("style", () => {
  test("Compile Style Base", () => {
    let styleContent = `
      .box{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });

  test("Compile Style With Scoped", () => {
    let styleContent = `
      .box{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: true,
      id: "test",
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });

  test("Compile Style With Multi Selector", () => {
    let styleContent = `
      .box.box-item{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });
  test("Compile Style With Multi Selector2", () => {
    let styleContent = `
      .box .box-item{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });
  test("Compile Style With Comma Selector", () => {
    let styleContent = `
      .box, .box-item{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });
  test("Compile Style With Comma Selector1(Less)", () => {
    let styleContent = `
      .box, 
      .box-item{
        box-sizing: border-box;
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });

  test("@media style", () => {
    // Expect Filter Media Info
    let styleContent = `
      @media screen and (min-width: 900px) {
        article {
          padding: 1rem 3rem;
        }
      }
    `;
    let code = compileStyle(styleContent, {
      scoped: false,
      packageName: "@hummer/tenon-vue",
    });
    expect(code).toMatchSnapshot();
  });
});
