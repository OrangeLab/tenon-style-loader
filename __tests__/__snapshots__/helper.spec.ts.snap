// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style @media style 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[]}};
    var options = {\\"scoped\\":false,\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;

exports[`style Compile Style Base 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[{\\"selector\\":\\"box\\",\\"n_selector\\":{\\"next\\":null,\\"value\\":\\"box\\",\\"matchType\\":0,\\"relation\\":0},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0}]}};
    var options = {\\"scoped\\":false,\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;

exports[`style Compile Style With Comma Selector 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[{\\"selector\\":\\"box\\",\\"n_selector\\":{\\"next\\":null,\\"value\\":\\"box\\",\\"matchType\\":0,\\"relation\\":0},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0},{\\"selector\\":\\"box-item\\",\\"n_selector\\":{\\"next\\":null,\\"value\\":\\"box-item\\",\\"matchType\\":0,\\"relation\\":0},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0}]}};
    var options = {\\"scoped\\":false,\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;

exports[`style Compile Style With Multi Selector 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[{\\"selector\\":\\"box-item\\",\\"n_selector\\":{\\"next\\":{\\"next\\":null,\\"value\\":\\"box\\",\\"matchType\\":0,\\"relation\\":0},\\"value\\":\\"box-item\\",\\"matchType\\":0,\\"relation\\":0},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0}]}};
    var options = {\\"scoped\\":false,\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;

exports[`style Compile Style With Multi Selector2 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[{\\"selector\\":\\"box-item\\",\\"n_selector\\":{\\"next\\":null,\\"value\\":\\"box-item\\",\\"matchType\\":0,\\"relation\\":1},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0}]}};
    var options = {\\"scoped\\":false,\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;

exports[`style Compile Style With Scoped 1`] = `
"
    import {collectStyle} from '@hummer/tenon-vue';
    export default (function(){
      
    var ruleSetMap = {\\"global\\":{\\"classList\\":[]},\\"test\\":{\\"classList\\":[{\\"selector\\":\\"box\\",\\"n_selector\\":{\\"next\\":null,\\"value\\":\\"box\\",\\"matchType\\":0,\\"relation\\":0},\\"style\\":{\\"boxSizing\\":\\"border-box\\"},\\"matchType\\":0}]}};
    var options = {\\"scoped\\":true,\\"id\\":\\"test\\",\\"packageName\\":\\"@hummer/tenon-vue\\"};
  
      return collectStyle(ruleSetMap, options);
    })();
  "
`;
